src/imageops/filter_1d.rs

```diff
@@ -0,0 +1,879 @@
+            let other_side = length - i - 1;
+            let fw = &src[(i * N)..(i * N) + 1];
+            let bw = &src[(other_side * N)..(other_side * N) + 1];
+            k0 = mla(k0, fw[0].as_() + bw[0].as_(), coeff);
+        }
+
+        *dst = k0.to_();
+    }
+}
+
+trait KernelTransformer<F, I> {
+    fn transform(input: F) -> I;
+}
+
+impl KernelTransformer<f32, u32> for u8 {
+    fn transform(input: f32) -> u32 {
+        const SCALE: f32 = (1 << Q0_15) as f32;
+        (input * SCALE).min(((1u32 << Q0_15) - 1) as f32).max(0.) as u32
+    }
+}
+
+impl KernelTransformer<f32, f32> for f32 {
+    fn transform(input: f32) -> f32 {
+        input
+    }
+}
+
+impl KernelTransformer<f32, f32> for u16 {
+    fn transform(input: f32) -> f32 {
+        input
+    }
+}
+
+/// Removes meaningless values from kernel preserving symmetry
+fn prepare_symmetric_kernel<I: Copy + PartialEq + 'static>(kernel: &[I]) -> Vec<I>
+where
+    i32: AsPrimitive<I>,
+{
+    let zeros: I = 0i32.as_();
+    let mut new_kernel = kernel.to_vec();
+    while new_kernel.len() > 2
+        && (new_kernel.last().unwrap().eq(&zeros) && new_kernel.first().unwrap().eq(&zeros))
+    {
+        new_kernel.remove(0);
+        new_kernel.remove(new_kernel.len() - 1);
+    }
+
+    new_kernel
+}
+
+const RING_QUEUE_CIRCULAR_CUTOFF: usize = 55;
+
+fn filter_1d_ring_queue<T, F, I, const N: usize>(
+    image: &[T],
+    destination: &mut [T],
+    image_size: FilterImageSize,
+    row_kernel: &[F],
+    column_kernel: &[F],
+) -> Result<(), ImageError>
+where
+    T: Copy + AsPrimitive<F> + Default + Send + Sync + KernelTransformer<F, I> + AsPrimitive<I>,
+    F: Default + 'static + Copy,
+    I: ToStorage<T>
+        + Mul<I, Output = I>
+        + Add<I, Output = I>
+        + MulAdd<I, Output = I>
+        + Send
+        + Sync
+        + PartialEq
+        + Default
+        + 'static
+        + Copy,
+    i32: AsPrimitive<F> + AsPrimitive<I>,
+    f64: AsPrimitive<T>,
+{
+    if image.len() != image_size.width * image_size.height * N {
+        return Err(ImageError::Limits(LimitError::from_kind(
+            LimitErrorKind::DimensionError,
+        )));
+    }
+    if destination.len() != image_size.width * image_size.height * N {
+        return Err(ImageError::Limits(LimitError::from_kind(
+            LimitErrorKind::DimensionError,
+        )));
+    }
+
+    let mut scanned_row_kernel = row_kernel
+        .iter()
+        .map(|&x| T::transform(x))
+        .collect::<Vec<I>>();
+    let mut scanned_column_kernel = column_kernel
+        .iter()
+        .map(|&x| T::transform(x))
+        .collect::<Vec<I>>();
+
+    scanned_row_kernel = prepare_symmetric_kernel(&scanned_row_kernel);
+    scanned_column_kernel = prepare_symmetric_kernel(&scanned_column_kernel);
+
+    if scanned_row_kernel.is_empty() || scanned_column_kernel.is_empty() {
+        for (dst, src) in destination.iter_mut().zip(image.iter()) {
+            *dst = *src;
+        }
```

Comment:
Since we have established the buffers are same length _and_ we have that `T: Copy` I'm wondering why this is not a `copy_from_slice` call?